v0.2.18: message UUID and component sequence on message metadata

matthiasn opened the pull request:
In this commit, UUIDs are added to the component metadata when a message is first
created. This allows for better tracking of the message traversing the system.
Also, the sequence of components handling a message is recorded.

The effects of this change can be seen in the example application, where some
messages routes are printed to the browser console or the terminal for the backend
application. I was undecided about the best way to record the sequence, so I created
two different formats. Try it out and let me know what you think. You'll need to
install the systems-toolbox locally for now as this WIP isn't published yet.

matthiasn committed:
v0.2.18: message UUID and component sequence on message metadata

In this commit, UUIDs are added to the component metadata when a message is first
created. This allows for better tracking of the message traversing the system.
Also, the sequence of components handling a message is recorded.

The effects of this change can be seen in the example application, where some
messages routes are printed to the browser console or the terminal for the backend
application. I was undecided about the best way to record the sequence, so I created
two different formats. Try it out and let me know what you think. You'll need to
install the systems-toolbox locally for now as this WIP isn't published yet.

matthiasn committed:
cleanup; correlation ID plus unique message ID

matthiasn committed:
UUID naming

kamituel commented on src/cljc/matthiasn/systems_toolbox/component.cljc:
```diff
@@ -33,6 +37,16 @@
+  thus far. Component IDs can be added either when the message enters or leaves a component.
+  There's a test to avoid conjoining the same ID twice in a row."
+  [msg-meta cmp-id]
+  (let [cmp-seq (vec (:cmp-seq msg-meta))]
```
Edge case: what if component will dispatch a message to itself?

kamituel replied:
I think it might be useful to only record new components when they enter. And then allow duplicates. I.e. if I misconfigure something, it'd be good to see that my message is trapped in a loop for instance.

matthiasn replied:
Hmm, does that really make sense for a component to send messages to itself? It can make sense to call another handler, but sending a message to itself here would mean a component's in channel is wired to the same component's out channel.

matthiasn replied:
Hadn't seen your second comment when commenting. But then we would miss the origin of the message, which is probably more important than some intermediate steps...

kamituel commented on src/cljc/matthiasn/systems_toolbox/component.cljc:
```diff
@@ -92,9 +106,11 @@
-                       msg-w-meta (with-meta msg msg-meta)]
+                                    (add-to-msg-seq cmp-id)
+                                    (assoc-in [cmp-id :out-ts] (now)))
+                       corr-id (uuid)
```
I'd add some note explaining that tag is a constant thing, while corr-id changes each time message is dispatches. It's clear now as we just discussed it, but might be useful 2 months from now ;)

matthiasn replied:
good idea, I'll do that

kamituel commented on changes.md:
````diff
@@ -1,3 +1,11 @@
+[matthiasn/systems-toolbox "0.2.18"]
+```
+
+* When a message is first emitted, a UUID is attached to the metadata, which allows tracking the message on its way through the system.
````
I'd say "a :tag (UUID) is attached to the metadata".
And also a sentence about the corr-id.

matthiasn committed:
addressing PR comments

matthiasn committed:
Change in cmp-seq recording behavior

Component IDs are either recorded when the cmp-seq is empty in the case of an
initial send or when the message is received by a component. This avoids
recording component IDs multiple times.
